style(vertical-nav): refactored css structure

mattnolting opened the pull request:
closes #1849 
closes #2196
closes #2650 

**Nav component preview**
https://patternfly-pr-2884.surge.sh/documentation/core/demos/page

## Breaking changes
The nav component's css was refactored, see the file list for details on what changed 
https://github.com/patternfly/patternfly/pull/2884/files

This PR:
1. Applies stylistic updates to css variables rather than redefining component elements.
2. Updates `--pf-c-nav__c-nested-component--Property` to `--pf-c-nav--c-nested-component--Property`. Any instances of the old variable should be updated to the new variable.
3. Updates visual style of navigation only, no HTML updates are necessary.

patternfly-build commented:
Preview: https://patternfly-pr-2884.surge.sh

christiemolloy commented:
Did design specify how the divider should look now that the items have borders ? @mcarrano 

<img width="433" alt="Screen Shot 2020-03-26 at 11 50 39 AM" src="https://user-images.githubusercontent.com/20118816/77667222-4b56b400-6f58-11ea-96b7-3313633e2984.png">

mcarrano commented:
Yeah, this does not look right to me @christiemolloy . Should we be having separators at all in the primary navigation? @mceledonia please also take a look at this. Looks good to me (other than the above issue), but would like you to review, also.

christiemolloy commented:
For the default nav, im looking at the mock ups, and it looks like the border/lines are the same color as the fill, which means that you wouldn't see the darker line:

<img width="839" alt="Screen Shot 2020-03-26 at 4 48 27 PM" src="https://user-images.githubusercontent.com/20118816/77695367-08a9d180-6f82-11ea-9a7b-7531ed7796a0.png">

<img width="328" alt="Screen Shot 2020-03-26 at 4 49 25 PM" src="https://user-images.githubusercontent.com/20118816/77695369-09426800-6f82-11ea-8700-eb2eb85e948a.png">

christiemolloy commented:
Looks like the subnav link should have a white left border on it

<img width="306" alt="Screen Shot 2020-03-26 at 4 53 53 PM" src="https://user-images.githubusercontent.com/20118816/77695607-69d1a500-6f82-11ea-86a8-1f602f1f26f1.png">

<img width="548" alt="Screen Shot 2020-03-26 at 4 53 50 PM" src="https://user-images.githubusercontent.com/20118816/77695603-68a07800-6f82-11ea-89aa-0cff21f364e7.png">

christiemolloy commented:
From the mock-ups im seeing this version without borders, are we supporting that ?

<img width="396" alt="Screen Shot 2020-03-26 at 4 56 55 PM" src="https://user-images.githubusercontent.com/20118816/77695901-e3699300-6f82-11ea-92a8-1bb33ec29842.png">

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -1,169 +1,118 @@
+  --pf-c-nav__list__link--PaddingTop: var(--pf-global--spacer--md);
+  --pf-c-nav__list__link--PaddingRight: var(--pf-global--spacer--lg);
+  --pf-c-nav__list__link--PaddingBottom: var(--pf-global--spacer--md);
+  --pf-c-nav__list__link--PaddingLeft: var(--pf-global--spacer--lg);
```
The horizontal padding for links should be `var(--pf-global--spacer--md)` < `$pf-global--breakpoint--md` to match the overall spacing changes that happen with the page layout between desktop and mobile. The left edge of the nav should align with the left edge of the hamburger on mobile.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -1,169 +1,118 @@
+  --pf-c-nav__simple-list__link--PaddingTop: var(--pf-global--spacer--sm);
+  --pf-c-nav__simple-list__link--PaddingRight: var(--pf-global--spacer--lg);
+  --pf-c-nav__simple-list__link--PaddingBottom: var(--pf-global--spacer--sm);
+  --pf-c-nav__simple-list__link--PaddingLeft: var(--pf-global--spacer--lg);
```
I imagine the simple-list__link padding will be the same - 16px horizontal on mobile, 24px on desktop

mcoker commented:
The `:active` and `:focus` styles on the horizontal navigation looks like this 

<img width="613" alt="Screen Shot 2020-03-26 at 6 37 19 PM" src="https://user-images.githubusercontent.com/35148959/77706632-eff4e800-6f90-11ea-92a2-7d300ce0638e.png">

christiemolloy commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -260,76 +222,12 @@
-    padding-left: var(--pf-c-nav__c-divider--MarginLeft);
-
-    // end - Update during breaking change
+    padding-right: var(--pf-c-nav--c-divider--PaddingRight);
```
why is padding being set to 0 here ? is it to override ? should it just be the value 0

mattnolting replied:
It's set once here so that it can inherit padding values in different contexts, or be themed.

christiemolloy commented:
This link doesn't appear to be disabled:

<img width="467" alt="Screen Shot 2020-03-26 at 7 52 58 PM" src="https://user-images.githubusercontent.com/20118816/77707423-757d9580-6f9b-11ea-9360-9b1c62daabe7.png">

mcoker commented:
An odd overflow issue with the horizontal nav

<img width="607" alt="Screen Shot 2020-03-26 at 7 02 07 PM" src="https://user-images.githubusercontent.com/35148959/77707834-54657680-6f94-11ea-81d2-20d92a8fef98.png">

mcoker commented:
Should tertiary nav have changed? The current component is blue on current/hover, and the text/underline are a darker shade of blue.

This PR
<img width="844" alt="Screen Shot 2020-03-26 at 7 10 16 PM" src="https://user-images.githubusercontent.com/35148959/77708211-74496a00-6f95-11ea-9703-a4d1bacabbc7.png">

Master/v4
<img width="808" alt="Screen Shot 2020-03-26 at 7 09 20 PM" src="https://user-images.githubusercontent.com/35148959/77708186-554ad800-6f95-11ea-8377-ef2a93e40072.png">

mcoker commented:
Looks like there is too much top padding on the horizontal nav between md and lg breakpoints

this pr
<img width="882" alt="Screen Shot 2020-03-26 at 7 15 52 PM" src="https://user-images.githubusercontent.com/35148959/77708511-4c0e3b00-6f96-11ea-842f-fe4cc59824b6.png">

master/v4
<img width="881" alt="Screen Shot 2020-03-26 at 7 15 58 PM" src="https://user-images.githubusercontent.com/35148959/77708519-503a5880-6f96-11ea-862d-a700d03f879e.png">

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,180 +252,105 @@
-  margin-bottom: var(--pf-c-nav__separator--MarginBottom);
-  margin-left: var(--pf-c-nav__separator--MarginLeft);
-  background-color: var(--pf-c-nav__separator--BackgroundColor);
+  &:active {
```
Not part of the design, but can you move anywhere we're defining `:active` along with `:focus` so that the `:active` styles come after? Otherwise they'll never work, since `:focus` will always override.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,180 +252,105 @@
+    top: 0;
+    right: 0;
+    bottom: 0;
+    left: 0;
```
This makes the text unselectable because it overlays the the entire link. Is there any reason you wouldn't use `width/height/background-color`?  That's typically how we create these treatments using pseudo elements.

mattnolting replied:
It's not because of the pseudo selectors. Disable them and try to select the text. It seems to be the caused by changing the `display` value and/or adding padding to an `<a>` element. Our current nav exhibits the same behavior.

mattnolting replied:
I find this method super efficient in that we don't have to specify multiple values to say, "Enable the lefthand border", like `top: 0; bottom: 0; left: 0; width: 4px; background-color: blue;`. Then for border right say `left: auto; right: 0; ` You can just say, "all border are 0 width and color blue, then give `--BorderLeftWidth` a width value to enable. It allows us to namespace active borders.

mcoker replied:
Yeah, I still think that as a general principle, we should not introduce an invisible overlay, just for styling purposes, over the element that keeps you from interacting with the things in the element under it. Unless it's necessary of course - though it doesn't seem necessary here. We define most borders using pseudo elements with `width/height/background-color` in other components, so it's also a more consistent approach as well.

mattnolting replied:
We'll need to update [tabs](https://patternfly-next-pr-2757.surge.sh/documentation/core/components/tabs) as well

mattnolting replied:
And looks like [buttons](https://pf4.patternfly.org/documentation/core/components/button)

https://github.com/patternfly/patternfly/issues/2893

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -230,12 +146,52 @@
+    }
   }
 
   // Disable underline
```
Not part of the redesign, but I think we could move this to `.pf-c-nav__link` and be a little more specific with the selector than `[class*="list"] [class*="link"]`

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -548,157 +372,91 @@
+// Section
 .pf-c-nav__section {
   margin-top: var(--pf-c-nav__section--MarginTop);
 
```
It's not in this changeset, but small nit - the line below this can be written as `+ &`

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -717,8 +475,26 @@
   font-size: var(--pf-c-nav__section-title--FontSize);
   font-weight: var(--pf-c-nav__section-title--FontWeight);
```
not part of the design, but looks like we can remove this prop & var.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -717,8 +475,26 @@
+  font-size: var(--pf-c-nav__toggle--FontSize);
+  line-height: 1;
+  pointer-events: none;
+  border: 0;
```
looks like these are legacy styles, but are `border` and `pointer-events` needed? Looks like it might have been a button at some point but isn't now as far as I can tell.

christiemolloy commented:
can you update the breaking change notes

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
-      background-color: var(--pf-c-nav__list-link--m-current--after--BackgroundColor);
-    }
+  &:hover {
+    color: var(--pf-c-nav__link--hover--Color);
```
should color and background-color redefine `--pf-c-nav__link--hover` and `--pf-c-nav__link--BackgroundColor` for these states?

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+.pf-c-nav__list,
+.pf-c-nav__simple-list {
+  .pf-c-nav__link.pf-m-current {
+    --pf-c-nav__link--before--BorderLeftWidth: var(--pf-c-nav__link--before--BorderWidth);
```
I wonder if we should separate these out so they aren't tied to one another? If we style them differently and `--pf-c-nav__link--before--BorderWidth` doesn't control both, that's a breaking change.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+  .pf-c-nav__link:focus,
+  .pf-c-nav__link:active,
+  .pf-c-nav__link.pf-m-current {
+    --pf-c-nav__link--before--BorderBottomWidth: var(--pf-c-nav__horizontal-list__link--before--BorderWidth);
```
Same here - what do you think about separating them out? It's less code to write this way and easier to theme, but less customizable and introduces a breaking change if we choose to style one differently. I think it's fair to expect we might style one differently as the simple nav does that.

mcoker replied:
Same with tertiary

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+  .pf-c-nav__link:not(.pf-m-current):hover {
+    --pf-c-nav__link--before--BorderLeftWidth: var(--pf-c-nav__subnav__link--not--m-current--hover--BorderLeftWidth);
+    --pf-c-nav__link--before--BorderColor: var(--pf-c-nav__subnav__link--not--m-current--hover--BorderColor);
+    --pf-c-nav__link--before--BorderTopWidth: 0;
```
can remove this

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -172,55 +114,27 @@
   // Scroll button
+  --pf-c-nav__scroll-button--Display: none;
+  --pf-c-nav__scroll-button--Visibility: hidden;
   --pf-c-nav__scroll-button--Top: var(--pf-global--spacer--sm);
```
looks like you can remove this

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -1,169 +1,111 @@
+  --pf-c-nav__link--BackgroundColor: transparent;
+  --pf-c-nav__link--hover--BackgroundColor: var(--pf-global--BackgroundColor--light-300);
+  --pf-c-nav__link--active--BackgroundColor: var(--pf-c-nav__link--hover--BackgroundColor);
+  --pf-c-nav__link--focus--BackgroundColor: var(--pf-c-nav__link--hover--BackgroundColor);
```
Same here - if we're going to have a var for each, we should probably not create the dependency between them if one changes.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+  --pf-c-nav__link--FontWeight: var(--pf-c-nav__tertiary-list__link--FontWeight);
+  --pf-c-nav__link--Color: var(--pf-c-nav__tertiary-list__link--Color);
+  --pf-c-nav__link--hover--Color: var(--pf-c-nav__tertiary-list__link--hover--Color);
+  --pf-c-nav__link--active--Color: var(--pf-c-nav__tertiary-list__link--hover--Color);
```
```suggestion
  --pf-c-nav__link--active--Color: var(--pf-c-nav__tertiary-list__link--active--Color);
```

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+  --pf-c-nav__link--Color: var(--pf-c-nav__tertiary-list__link--Color);
+  --pf-c-nav__link--hover--Color: var(--pf-c-nav__tertiary-list__link--hover--Color);
+  --pf-c-nav__link--active--Color: var(--pf-c-nav__tertiary-list__link--hover--Color);
+  --pf-c-nav__link--focus--Color: var(--pf-c-nav__tertiary-list__link--hover--Color);
```
```suggestion
  --pf-c-nav__link--focus--Color: var(--pf-c-nav__tertiary-list__link--focus--Color);
```

mcoker commented on src/patternfly/demos/Page/page-demo-expandable-nav.hbs:
```diff
@@ -84,11 +84,6 @@
-          {{#> nav-separator}}{{/nav-separator}}
-        {{else}}
-          {{#> divider divider--type="li"}}{{/divider}}
-        {{/if}}
```
Did you want to pull the nav-separator out of this component in this PR? Works for me if you do, I saw this and removed that element from the documentation, but realized we still need to yank out

https://github.com/mattnolting/patternfly/blob/3fa5a9097a7487fc87d84f776ea60218b1ef07d3/src/patternfly/demos/Page/page-demo-expandable-nav.hbs#L45-L49

https://github.com/mattnolting/patternfly/blob/style-vertical-nav/src/patternfly/components/Nav/nav-separator.hbs

and the 2 examples in https://github.com/mattnolting/patternfly/blob/style-vertical-nav/src/patternfly/demos/Page/examples/Page.md that reference `IsLegacySeparators`

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+  --pf-c-nav__link--hover--BackgroundColor: var(--pf-c-nav__tertiary-list__link--hover--BackgroundColor);
+  --pf-c-nav__link--active--BackgroundColor: var(--pf-c-nav__tertiary-list__link--hover--BackgroundColor);
+  --pf-c-nav__link--focus--BackgroundColor: var(--pf-c-nav__tertiary-list__link--hover--BackgroundColor);
+  --pf-c-nav__link--m-current--BackgroundColor: var(--pf-c-nav__tertiary-list__link--m-current--BackgroundColor);
```
`--pf-c-nav__tertiary-list__link--m-current--BackgroundColor` is undefined.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +235,275 @@
+  --pf-c-nav__link--hover--BackgroundColor: var(--pf-c-nav__horizontal-list__link--hover--BackgroundColor);
+  --pf-c-nav__link--active--BackgroundColor: var(--pf-c-nav__horizontal-list__link--hover--BackgroundColor);
+  --pf-c-nav__link--focus--BackgroundColor: var(--pf-c-nav__horizontal-list__link--hover--BackgroundColor);
+  --pf-c-nav__link--m-current--BackgroundColor: var(--pf-c-nav__horizontal-list__link--m-current--BackgroundColor);
```
`--pf-c-nav__horizontal-list__link--m-current--BackgroundColor` is undefined. Another approach could be to set the following to transparent by default, then only declare them where they're needed, instead of having to unset them all in horizontal and tertiary nav

```
  --pf-c-nav__link--BackgroundColor
  --pf-c-nav__link--hover--BackgroundColor
  --pf-c-nav__link--active--BackgroundColor
  --pf-c-nav__link--focus--BackgroundColor
  --pf-c-nav__link--m-current--BackgroundColor
```

mattnolting commented:
@mcoker updated!

mcoker commented:
Doesn't look like the `pf-m-start/end` classes are working on the horizontal nav:

<img width="1352" alt="Screen Shot 2020-04-03 at 2 18 20 PM" src="https://user-images.githubusercontent.com/35148959/78397015-095fea80-75b6-11ea-9090-474bacd585c0.png">
<img width="1316" alt="Screen Shot 2020-04-03 at 2 19 37 PM" src="https://user-images.githubusercontent.com/35148959/78397065-2d233080-75b6-11ea-9d1f-6f44bc4b3121.png">

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +253,290 @@
-
-  // Expanded list item
   .pf-c-nav__item.pf-m-expanded & {
     max-height: var(--pf-c-nav__subnav--MaxHeight);
```
This var name is probably better as `--pf-c-nav__item--m-expanded__subnav--MaxHeight` since `__subnav` actually has a `max-height`

And since you declare `max-height` for `__subnav`, you could use `--pf-c-nav__subnav--MaxHeight` there, then here declare `--pf-c-nav__subnav--MaxHeight: var(--pf-c-nav__item--m-expanded__subnav--MaxHeight);`

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +253,290 @@
 
-  + .pf-c-nav__section {
+  + & {
     --pf-c-nav__section--MarginTop: var(--pf-c-nav__section__section--MarginTop);
```
can we change this to `--pf-c-nav__section--section--MarginTop`?

christiemolloy commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +251,290 @@
+
+  .pf-c-nav__link:not(.pf-m-current):hover {
+    --pf-c-nav__link--before--BorderLeftWidth: var(--pf-c-nav__subnav__link--not--m-current--hover--BorderLeftWidth);
+    --pf-c-nav__link--before--BorderColor: var(--pf-c-nav__subnav__link--not--m-current--hover--BorderColor);
```
could this variable just be `--pf-c-nav__subnav__link--hover--BorderColor`, since that is the main case

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -462,20 +462,21 @@
   --pf-c-nav__simple-list__link--FontSize: var(--pf-c-nav__subnav__link--FontSize);
 
-  max-height: 0;
+  max-height: var(--pf-c-nav__subnav--MaxHeight);
```
looks like we need to define `--pf-c-nav__subnav--MaxHeight: 0;`

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -474,9 +485,24 @@
-  .pf-c-nav__link:not(.pf-m-current):hover {
-    --pf-c-nav__link--before--BorderLeftWidth: var(--pf-c-nav__subnav__link--hover--BorderLeftWidth);
-    --pf-c-nav__link--before--BorderColor: var(--pf-c-nav__subnav__link--hover--BorderColor);
+  .pf-c-nav__link:hover {
```
very nice!

mattnolting commented:
@mcarrano @mceledonia updated per our meeting

- Removed references in examples
- Removed `disabled` background color in light theme

@mcoker noted that this fails color contrast
https://files.slack.com/files-pri/T0FA4HM8D-F011AUCHD8S/screen_shot_2020-04-07_at_12.52.12_pm.png

I also left the cursor as is right now. We can open another issue to track.

mcoker commented:
Looks like the hover state for the scroll buttons changed colors - is that intended?

<img width="193" alt="Screen Shot 2020-04-07 at 1 19 09 PM" src="https://user-images.githubusercontent.com/35148959/78704999-68817000-78d2-11ea-960d-7eaab01d555a.png">

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +268,330 @@
-        background-color: var(--pf-c-nav__horizontal-list-link--focus--after--BackgroundColor);
-      }
-    }
+  .pf-c-nav__link::after {
```
can you nest this and the `::before` in the `.pf-c-nav__link` selector above?

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +268,330 @@
-      height: var(--pf-c-nav__horizontal-list-link--after--Height);
-      content: "";
-    }
+  .pf-c-nav__link.pf-m-current {
```
can you scss nest hover/focus/active/current?

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -354,389 +268,330 @@
-        background-color: var(--pf-c-nav__tertiary-list-link--focus--after--BackgroundColor);
-      }
-    }
+  .pf-c-nav__link.pf-m-current {
```
can you nest hover/focus/active/current?

mcoker commented:
@mattnolting for markup changes in the breaking changes notes, can you verify that the nav separator was removed in this PR and the divider should be used instead? Also looks like this is where there was a change to the `.pf-c-nav__toggle-icon` element applied to a wrapper instead of the icon.
